client-common/src/tendermint/websocket_rpc_client/websocket_rpc_loop.rs

```diff
@@ -0,0 +1,229 @@
+#![cfg(feature = "websocket-rpc")]
+use std::collections::HashMap;
+use std::sync::mpsc::SyncSender;
+use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+
+use websocket::receiver::Reader;
+use websocket::sender::Writer;
+use websocket::stream::sync::TcpStream;
+use websocket::{ClientBuilder, OwnedMessage};
+
+use crate::{ErrorKind, Result, ResultExt};
+
+use super::{ConnectionState, JsonRpcResponse};
+
+const MONITOR_RETRY_INTERVAL: Duration = Duration::from_secs(2);
+
+/// Creates a new websocket connection with given url
+pub fn new_connection(url: &str) -> Result<(Reader<TcpStream>, Writer<TcpStream>)> {
+    ClientBuilder::new(url)
+        .chain(|| (ErrorKind::InvalidInput, format!("Malformed url: {}", url)))?
+        .connect_insecure()
+        .chain(|| {
+            (
+                ErrorKind::InitializationError,
+                format!("Unable to connect to websocket RPC at: {}", url),
+            )
+        })?
+        .split()
+        .chain(|| {
+            (
+                ErrorKind::InternalError,
+                "Unable to split websocket reader and writer",
+            )
+        })
+}
+
+/// Spawns websocket rpc loop in a new thread
+///
+/// # How it works
+///
+/// - Connects to websocket server at given `url` and splits the connection in `reader` and `writer`.
+/// - Spawns a thread and runs `websocket_rpc_loop` in the thread which continues until the thread panics.
+/// - For each websocket message received:
+///   - Parse the message into JSON-RPC response.
+///   - Pop the response channel from `channel_map` corresponding to response's `request_id`.
+///   - Send the response to the channel.
+#[allow(clippy::type_complexity)]
```

Comment:
what flags the type complexity? the channel map?

Reply:
Nothing. I added this by earlier and modified signature after that. Removed.